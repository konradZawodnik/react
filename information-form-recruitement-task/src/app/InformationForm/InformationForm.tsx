"use client";
import React, { useState, useCallback, useMemo, useEffect } from "react";
import { useForm, SubmitHandler } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { Container, Row, Col } from "react-bootstrap";
import { DraggableBullet } from "./DraggableBullet/DraggableBullet";
import { z } from "zod";
import { useDrop } from "react-dnd";
import CreatableSelect from "react-select/creatable";
import dynamic from "next/dynamic";
import { Form } from "@/components/ui/form";
import {
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import "react-quill/dist/quill.snow.css";

export const ItemType = "BULLET";

export default function InformationForm() {
  const [bullets, setBullets] = useState<string[]>([]);

  const FormSchema = z.object({
    title: z.string({
      required_error: "Product title is required",
    }),
    description: z.string().optional(),
    bullets: z.array(z.string()).optional(),
    keywords: z.array(z.string()).optional(),
  });

  type FormData = z.infer<typeof FormSchema>;

  const form = useForm<z.infer<typeof FormSchema>>({
    resolver: zodResolver(FormSchema),
    mode: "onChange",
  });
  const {
    control,
    handleSubmit,
    register,
    setValue,
    watch,
  } = form;

  const ReactQuill = useMemo(
    () => dynamic(() => import("react-quill"), { ssr: false }),
    []
  );

  const onSubmit: SubmitHandler<FormData> = (data) => {
    console.log(data);
  };

  const moveBullet = useCallback(
    (dragIndex: number, hoverIndex: number) => {
      const updatedBullets = [...bullets];
      const [draggedBullet] = updatedBullets.splice(dragIndex, 1);
      updatedBullets.splice(hoverIndex, 0, draggedBullet);
      setBullets(updatedBullets);
      setValue("bullets", updatedBullets);
    },
    [bullets, setValue]
  );

  const [, drop] = useDrop({
    accept: ItemType,
    drop: (item: { index: number }) => moveBullet(item.index, bullets.length),
    collect: (monitor) => ({
      isOver: !!monitor.isOver(),
    }),
  });

  const handleAddBullet = () => {
    const newBullets = [...bullets, ""];
    setBullets(newBullets);
    setValue("bullets", newBullets);
  };

  const handleRemoveBullet = (index: number) => {
    const updatedBullets = bullets.filter((_, i) => i !== index);
    setBullets(updatedBullets);
    setValue("bullets", updatedBullets);
  };

  const handleChangeBullet = (index: number, value: string) => {
    const updatedBullets = [...bullets];
    updatedBullets[index] = value;
    setBullets(updatedBullets);
    setValue("bullets", updatedBullets);
  };

  const handleKeywordChange = (newValue: any) => {
    setValue(
      "keywords",
      newValue ? newValue.map((item: { label: string }) => item.label) : []
    );
  };

  useEffect(() => {
    register("description", { required: false });
  }, [register]);

  const onEditorStateChange = (editorState: string) => {
    setValue("description", editorState);
  };

  const editorContent = watch("description");

  return (
    <Container>
      <Row>
        <Col md={8}>
          <Form {...form}>
            <Label className="flex font-bold justify-center mb-4 mt-3 text-lg">
              Product Information Form
            </Label>
            <form onSubmit={handleSubmit(onSubmit)}>
              <div className="mb-4 space-y-4 w-full">
                <FormField
                  control={control}
                  name="title"
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel className="flex justify-center text-base text-[#112950]">
                        Product Title
                      </FormLabel>
                      <FormControl>
                        <Input
                          {...field}
                          className="border border-gray-500 rounded"
                          placeholder="Enter product title"
                          type="text"
                          value={field.value}
                        />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
              </div>

              <div className="flex flex-col mb-4 rounded">
                <FormField
                  control={control}
                  name="description"
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel className="flex justify-center mb-2 text-base">
                        Product Description
                      </FormLabel>
                      <FormControl>
                        <ReactQuill
                          {...field}
                          className="h-full"
                          theme="snow"
                          value={editorContent}
                          onChange={onEditorStateChange}
                        />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
              </div>

              <div
                className={`flex flex-col ${
                  bullets.length ? "mb-[calc(12vh)]" : "mb-2"
                } max-h-32 rounded`}
              >
                <FormField
                  name="bullets"
                  control={control}
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel className="flex justify-center mb-2 text-base">
                        Product Bullets
                      </FormLabel>
                      <FormControl>
                        <div
                          className="h-full max-h-[150px] border border-gray-500 mb-4 overflow-auto p-2.5"
                          ref={(node) => {
                            drop(node);
                          }}
                        >
                          {bullets.map((bullet, index) => (
                            <DraggableBullet
                              {...field}
                              key={index}
                              index={index}
                              bullet={bullet}
                              moveBullet={moveBullet}
                              handleRemoveBullet={handleRemoveBullet}
                              handleChangeBullet={handleChangeBullet}
                            />
                          ))}
                          <Button
                            className="bg-[#0068fa] border border-solid cursor-pointer hover:bg-[#0068fa] p-2 rounded text-white w-full"
                            variant="default"
                            onClick={(e) => {
                              e.preventDefault();
                              handleAddBullet();
                            }}
                          >
                            Add Bullet
                          </Button>
                        </div>
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
              </div>

              <div className="flex flex-col mb-4">
                <FormField
                  name="keywords"
                  control={control}
                  render={({ field }) => (
                    <FormItem>
                      <FormLabel
                        aria-label="Product Keywords"
                        className="flex justify-center mb-2 text-base"
                      >
                        Product Keywords
                      </FormLabel>
                      <FormControl>
                        <CreatableSelect
                          {...field}
                          className="cursor-pointer"
                          isMulti
                          onChange={handleKeywordChange}
                          options={[]}
                          value={field.value?.map((keyword) => ({
                            label: keyword,
                            value: keyword,
                          }))}
                          isClearable
                        />
                      </FormControl>
                      <FormMessage />
                    </FormItem>
                  )}
                />
              </div>

              <div className="flex flex-col justify-center items-center mb-4 w-full">
                <Button
                  className="bg-[#0068fa] border border-solid cursor-pointer flex justify-center hover:bg-[#0068fa] max-w-32 ml-0 mt-4 p-2 rounded text-white w-full"
                  type="submit"
                >
                  Submit
                </Button>
              </div>
            </form>
          </Form>
        </Col>
      </Row>
    </Container>
  );
}
