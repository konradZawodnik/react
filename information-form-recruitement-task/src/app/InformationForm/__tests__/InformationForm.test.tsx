import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import InformationForm from "../InformationForm";
import "@testing-library/jest-dom";

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

jest.mock("react-quill", () => {
  const MockComponent = () => <div data-testid="react-quill" />;
  MockComponent.displayName = "ReactQuill";
  return MockComponent;
});

jest.mock("react-dnd", () => ({
  useDrag: jest
    .fn()
    .mockImplementation(() => [jest.fn(), jest.fn(), jest.fn()]),
  useDrop: jest
    .fn()
    .mockImplementation(() => [jest.fn(), jest.fn(), jest.fn()]),
}));

describe("InformationForm", () => {
  it("should match snapshot", () => {
    const { container } = render(<InformationForm />);
    expect(container).toMatchSnapshot();
  });
  
  it("should render the form correctly", () => {
    render(<InformationForm />);

    expect(
      screen.getByPlaceholderText(/Enter product title/i)
    ).toBeInTheDocument();
    expect(screen.getByText(/Product Description/i)).toBeInTheDocument();
    expect(screen.getByText(/Product Bullets/i)).toBeInTheDocument();
    expect(screen.getByText(/Product Keywords/i)).toBeInTheDocument();
    expect(screen.getByText(/Submit/i)).toBeInTheDocument();
  });

  it("should add a bullet point", async () => {
    render(<InformationForm />);

    const addBulletButton = screen.getByText("Add Bullet");
    await userEvent.click(addBulletButton);

    expect(screen.getByPlaceholderText("Bullet point #1")).toBeInTheDocument();
  });

  it("should modify a bullet point", async () => {
    render(<InformationForm />);

    const addBulletButton = screen.getByText("Add Bullet");
    await userEvent.click(addBulletButton);

    const bulletInput = screen.getByPlaceholderText("Bullet point #1");
    await userEvent.type(bulletInput, "New Bullet Point");

    expect(bulletInput).toHaveValue("New Bullet Point");
  });

  it("should remove a bullet point", async () => {
    render(<InformationForm />);

    const addBulletButton = screen.getByText("Add Bullet");
    await userEvent.click(addBulletButton);

    const removeButton = screen.getByText("Remove");
    await userEvent.click(removeButton);

    await waitFor(() => {
      expect(
        screen.queryByPlaceholderText("Bullet point #1")
      ).not.toBeInTheDocument();
    });
  });

  it("should submit the form", async () => {
    render(<InformationForm />);
    const addBulletButton = screen.getByText("Add Bullet");
    const consoleLogSpy = jest.spyOn(console, "log");

    await userEvent.type(
      screen.getByPlaceholderText("Enter product title"),
      "Test Product"
    );
    await userEvent.click(addBulletButton);
    await userEvent.type(
      screen.getByPlaceholderText("Bullet point #1"),
      "Bullet 1"
    );
    await userEvent.click(addBulletButton);
    await userEvent.type(
      screen.getByPlaceholderText("Bullet point #2"),
      "Bullet 2"
    );

    const submitButton = screen.getByText("Submit");
    await waitFor(() => userEvent.click(submitButton));

    await waitFor(() => {
      expect(consoleLogSpy).toHaveBeenCalledWith({
        bullets: ["Bullet 1", "Bullet 2"],
        description: undefined,
        keywords: undefined,
        title: "Test Product",
      });
      consoleLogSpy.mockRestore();
    });
  });

  it("should allow creating, selecting and removing options in CreatableSelect", async () => {
    render(<InformationForm />);

    const creatableSelect = screen.getByText(/Select.../i);
    expect(creatableSelect).toBeInTheDocument();

    await userEvent.type(creatableSelect, "New Keyword{enter}");

    const keywordElement = await screen.findByText("New Keyword");
    expect(keywordElement).toBeInTheDocument();

    const removeKeywordButton = screen
      .getByText("New Keyword").nextElementSibling;

    if (removeKeywordButton) {
      await userEvent.click(removeKeywordButton);
      expect(screen.queryByText("New Keyword")).not.toBeInTheDocument();
    } else {
      throw new Error("Remove button not found");
    }
  });
});
