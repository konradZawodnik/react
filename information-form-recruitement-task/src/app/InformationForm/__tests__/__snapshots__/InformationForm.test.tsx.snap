// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InformationForm should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-8"
      >
        <label
          class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex font-bold justify-center mb-4 mt-3 text-lg"
        >
          Product Information Form
        </label>
        <form>
          <div
            class="mb-4 space-y-4 w-full"
          >
            <div
              class="space-y-2"
            >
              <label
                class="font-medium peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex justify-center text-base text-[#112950]"
                for=":r0:-form-item"
              >
                Product Title
              </label>
              <input
                aria-describedby=":r0:-form-item-description"
                aria-invalid="false"
                class="flex h-9 w-full bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm border border-gray-500 rounded"
                id=":r0:-form-item"
                name="title"
                placeholder="Enter product title"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="flex flex-col mb-4 rounded"
          >
            <div
              class="space-y-2"
            >
              <label
                class="font-medium peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex justify-center mb-2 text-base"
                for=":r1:-form-item"
              >
                Product Description
              </label>
            </div>
          </div>
          <div
            class="flex flex-col mb-2 max-h-32 rounded"
          >
            <div
              class="space-y-2"
            >
              <label
                class="font-medium peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex justify-center mb-2 text-base"
                for=":r2:-form-item"
              >
                Product Bullets
              </label>
              <div
                aria-describedby=":r2:-form-item-description"
                aria-invalid="false"
                class="h-full max-h-[150px] border border-gray-500 mb-4 overflow-auto p-2.5"
                id=":r2:-form-item"
              >
                <button
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 bg-[#0068fa] border border-solid cursor-pointer hover:bg-[#0068fa] p-2 rounded text-white w-full"
                >
                  Add Bullet
                </button>
              </div>
            </div>
          </div>
          <div
            class="flex flex-col mb-4"
          >
            <div
              class="space-y-2"
            >
              <label
                aria-label="Product Keywords"
                class="font-medium peer-disabled:cursor-not-allowed peer-disabled:opacity-70 flex justify-center mb-2 text-base"
                for=":r3:-form-item"
              >
                Product Keywords
              </label>
              <div
                class="cursor-pointer css-b62m3t-container"
                id=":r3:-form-item"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-2-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                  role="log"
                />
                <div
                  class="css-13cymwt-control"
                >
                  <div
                    class="css-hlgwow"
                  >
                    <div
                      class="css-1jqq78o-placeholder"
                      id="react-select-2-placeholder"
                    >
                      Select...
                    </div>
                    <div
                      class="css-19bb58m"
                      data-value=""
                    >
                      <input
                        aria-activedescendant=""
                        aria-autocomplete="list"
                        aria-describedby="react-select-2-placeholder"
                        aria-expanded="false"
                        aria-haspopup="true"
                        aria-invalid="false"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class=""
                        id="react-select-2-input"
                        role="combobox"
                        spellcheck="false"
                        style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="css-1wy0on6"
                  >
                    <span
                      class="css-1u9des2-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="css-1xc3v61-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div>
                  <input
                    name="keywords"
                    type="hidden"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="flex flex-col justify-center items-center mb-4 w-full"
          >
            <button
              class="items-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 bg-[#0068fa] border border-solid cursor-pointer flex justify-center hover:bg-[#0068fa] max-w-32 ml-0 mt-4 p-2 rounded text-white w-full"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
